src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs

```diff
@@ -421,20 +433,34 @@ public IEnumerable<HdAddress> GetUnusedAddresses(WalletAccountReference accountR
                 // Get the account.
                 HdAccount account = wallet.GetAccountByCoinType(accountReference.AccountName, this.coinType);
 
-                List<HdAddress> unusedAddresses = account.ExternalAddresses.Where(acc => !acc.Transactions.Any()).ToList();
+                List<HdAddress> unusedAddresses = isChange ? 
+                    account.InternalAddresses.Where(acc => !acc.Transactions.Any()).ToList() : 
+                    account.ExternalAddresses.Where(acc => !acc.Transactions.Any()).ToList();
+                
                 int diff = unusedAddresses.Count - count;
                 if (diff < 0)
                 {
-                    IEnumerable<HdAddress> newReceivingAddresses = account.CreateAddresses(this.network, Math.Abs(diff), isChange: false);
-                    this.UpdateKeysLookupLock(newReceivingAddresses);
+                    IEnumerable<HdAddress> newAddresses = account.CreateAddresses(this.network, Math.Abs(diff), isChange: isChange);
+                    this.UpdateKeysLookupLock(newAddresses);
                     generated = true;
                 }
 
-                addresses = account
-                    .ExternalAddresses
-                    .Where(acc => !acc.Transactions.Any())
-                    .OrderBy(x => x.Index)
-                    .Take(count);
+                if (isChange)
+                {
```

Comment:
you already did this enumeration at line 436 why do it again?

Reply:
I guess yeah. I didn't do it because if say you have 7 unused addresses and you want to receive 3, unusedAddresses will return all 7 addresses.
But you can return only the first 3 (.Take(3)). I'll change it.